File: python/037.py
#!/usr/bin/env python3
"""
Project Euler Problem 37: Truncatable primes

Find the sum of the only eleven primes that are both truncatable from left to right and right to left.
"""

from math import sqrt


def is_prime(n: int) -> bool:
    """Check if a number is prime."""
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True


def is_left_truncatable(n: int) -> bool:
    """Check if prime is left truncatable."""
    if n < 10:
        return False
    s = str(n)
    for i in range(1, len(s)):
        if not is_prime(int(s[i:])):
            return False
    return True


def is_right_truncatable(n: int) -> bool:
    """Check if prime is right truncatable."""
    if n < 10:
        return False
    s = str(n)
    for i in range(1, len(s)):
        if not is_prime(int(s[:-i])):
            return False
    return True


def main():
    truncatable_primes = []
    num = 11
    
    while len(truncatable_primes) < 11:
        if is_prime(num) and is_left_truncatable(num) and is_right_truncatable(num):
            truncatable_primes.append(num)
        num += 2
        if num % 5 == 0:
            num += 2
    
    print(sum(truncatable_primes))


if __name__ == "__main__":
    main()

File: c/037.c
/*
 * Project Euler 037 - Truncatable Primes
 * Find the sum of the only eleven primes that are both truncatable
 * from left to right and right to left.
 */
#include <stdio.h>
#include <stdbool.h>

bool is_prime(int n) {
    if (n < 2) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;
    for (int i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}

/* Count digits */
int num_digits(int n) {
    int count = 0;
    while (n > 0) { count++; n /= 10; }
    return count;
}

/* Compute 10^n */
int power10(int n) {
    int r = 1;
    for (int i = 0; i < n; i++) r *= 10;
    return r;
}

bool is_left_truncatable(int n) {
    /* n is already known prime and > 10 */
    int len = num_digits(n);
    for (int i = 1; i < len; i++) {
        int truncated = n % power10(i);
        if (!is_prime(truncated)) return false;
    }
    return true;
}

bool is_right_truncatable(int n) {
    /* n is already known prime and > 10 */
    int temp = n / 10;
    while (temp > 0) {
        if (!is_prime(temp)) return false;
        temp /= 10;
    }
    return true;
}

int main(void) {
    int count = 0;
    int sum = 0;
    int num = 11;

    while (count < 11) {
        if (is_prime(num) && is_left_truncatable(num) && is_right_truncatable(num)) {
            sum += num;
            count++;
        }
        num += 2;
        if (num % 5 == 0) num += 2;
    }

    printf("%d\n", sum);
    return 0;
}
